spring-integration-file/src/main/java/org/springframework/integration/file/remote/RemoteFileTemplate.java

```diff
@@ -434,6 +453,28 @@ public boolean get(final String remotePath, final InputStreamCallback callback)
 	}
 
 	@Override
+	public <T> T invoke(OperationsCallback<F, T> action) {
+		Session<F> contextSession = this.contextSessions.get();
+		if (contextSession == null) {
+			this.contextSessions.set(this.sessionFactory.getSession());
+		}
+		this.activeTemplateCallbacks.incrementAndGet();
```

Comment:
I leave the logic in this method as is. That isn't so hard to track recursion here. Plus check our `MPUT` impl, especially when it is recursive as well. In the end it call `doPut()` for each file, which is based on the `invoke()` as well, since we have there not only `send()` operation, but `doChmod()`, too.